### DISCUSSION TITLE
Redirect after Oauth login

### DISCUSSION THREAD

**Discussion by bert2002** [fastapi-users/fastapi-users#1173] [Q&A] [answered]
Hi,
I implemented a simple oauth login with Google and after the authentication the user gets redirected back to e.g.: `http://127.0.0.1:8000/auth/google/callback?state=eyJhb...`. Since I am using CookieTransport the frontend application does not need to get a JWT token. 

What is the best way to add a server side redirect after a callback? I was thinking of `on_after_login` in the `UserManager`, but it does not invoke anything when I use a return.

Thanks,
bert

**Comment by frankie567** [comment, marked as answer]
What you can do is to overload `CookieTransport` and customize `get_login_response` method:

```py
class CookieRedirectTransport(CookieTransport):
    async def get_login_response(self, token: str, response: Response) -> Any:
        await super().get_login_response(token, response)
        response.status_code = 302
        response.headers["Location"] = "http://localhost:3000/your-redirect"
```

and use this class as your transport

**Reply by bert2002** [reply]
That works perfectly :)

**Reply by frankie567** [reply]
Great to hear!

(took the liberty to reopen the discussion ; I'm not really a fan of this new feature, I think questions should remain visible for everyone to see 😄)

**Reply by madeinoz67** [reply]
thanks was exactly what I was looking for

**Comment by FullPortDev** [comment]
for future lost souls.

```py
class CookieRedirectTransport(CookieTransport):
    async def get_login_response(self, token: str) -> Any:
        response = RedirectResponse("http://localhost:5173/success", 302)
        self._set_login_cookie(response, token)
        return response

```

**Comment by bert2002** [comment]
I just updated to version `v11.0.0` and I am now getting

```
  File "/usr/local/lib/python3.10/site-packages/fastapi_users/router/oauth.py", line 150, in callback
    response = await backend.login(strategy, user)
  File "/usr/local/lib/python3.10/site-packages/fastapi_users/authentication/backend.py", line 46, in login
    return await self.transport.get_login_response(token)
TypeError: CookieRedirectTransport.get_login_response() missing 1 required positional argument: 'response'
```
Then I looked at the release notes and updated `on_after_login`, but no success. Further looking into it I saw @FullPortDev solution to return the `response`, but when I am using this, the user is not logged in after the authentication (even so it works). Anything else that needs to be updated?

**Reply by bert2002** [reply]
I tried different ways, but without success. Is it a bug or I just do something wrong?

**Reply by FullPortDev** [reply]
there's not enough info to fully answer, but here's something to check

# 1 are you including creds in your calls?
fetch({
credentials: 'include',
})

# 2 are the cookies added to the browser? 
dev console > application > cookies

# 3 are the sessions being added to your strategy db? 
- is the strategy setup properly

# 4 do you setup the backend like the following
```
auth_backend = AuthenticationBackend(
    name="cookie",
    transport=cookie_transport,
    get_strategy=get_redis_strategy,
)
```

**Reply by bert2002** [reply]
@FullPortDev I will be damned, but your code works now 🤔
